domain/network/state/space.go

```diff
@@ -0,0 +1,250 @@
+// Copyright 2023 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package state
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+
+	"github.com/canonical/sqlair"
+	"github.com/juju/errors"
+	"github.com/juju/names/v4"
+	"github.com/juju/utils/v3"
+
+	coreDB "github.com/juju/juju/core/database"
+	"github.com/juju/juju/core/network"
+	"github.com/juju/juju/domain"
+)
+
+// State represents a type for interacting with the underlying state.
+type State struct {
+	*domain.StateBase
+}
+
+// NewState returns a new State for interacting with the underlying state.
+func NewState(factory coreDB.TxnRunnerFactory) *State {
+	return &State{
+		StateBase: domain.NewStateBase(factory),
+	}
+}
+
+// AddSpace creates and returns a new space.
+func (st *State) AddSpace(
+	ctx context.Context,
+	uuid utils.UUID,
+	name string,
+	providerID network.Id,
+	subnetIDs []string,
+) error {
+	if !names.IsValidSpace(name) {
+		return errors.NewNotValid(nil, fmt.Sprintf("invalid space name '%s'", name))
+	}
+
+	db, err := st.DB()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	insertSpaceStmt := `
+INSERT INTO space (uuid, name)
+VALUES (?, ?)`
+	insertProviderStmt := `
+INSERT INTO provider_space (provider_id, space_uuid)
+VALUES (?, ?)`
+	checkSubnetFanLocalUnderlay := `
+SELECT subnet.cidr,subnet_type.is_space_settable
+FROM   subnet_type
+JOIN   subnet
+ON     subnet.subnet_type_id = subnet_type.id
+WHERE  subnet.uuid = ?`
+	err = db.StdTxn(ctx, func(ctx context.Context, tx *sql.Tx) error {
+		if _, err := tx.ExecContext(ctx, insertSpaceStmt, uuid.String(), name); err != nil {
+			return errors.Trace(err)
+		}
+		if providerID != "" {
+
+			if _, err := tx.ExecContext(ctx, insertProviderStmt, providerID, uuid.String()); err != nil {
+				return errors.Trace(err)
+			}
+		}
+		for _, subnetID := range subnetIDs {
+			// Check if the subnet is a fan overlay, in that case
+			// the space cannot be created on this subnet: it must
+			// be inherited from the underlay.
+			var (
+				cidr            string
+				isSpaceSettable bool
+			)
+			row := tx.QueryRowContext(ctx, checkSubnetFanLocalUnderlay, subnetID)
+			if err := row.Scan(&cidr, &isSpaceSettable); err != nil {
+				return errors.Trace(err)
+			}
+			if !isSpaceSettable {
+				return errors.Errorf(
+					"cannot set space for FAN subnet %q - it is always inherited from underlay", cidr)
+			}
+
+			if err := updateSubnetSpaceIDTx(ctx, tx, subnetID, uuid.String()); err != nil {
+				return errors.Trace(err)
+			}
+		}
+		return nil
+	})
+	return errors.Trace(err)
+}
+
+const retrieveSpacesStmt = `
+SELECT     
```

Comment:
we cannot add a space over a fan subnet, right? So why should we join over the fan subnets since the space should never be linked to those subnets?

Reply:
Because the Fan overlays are implicitly in the same space as the underlay. So we need to include them in the answer to the question "what subnets are in this space?"

Reply:
done, see e87b47feab